File: python/331_VerifyPreorderSerializationofaBinaryTree.py
from unittest import TestCase
# https://leetcode.com/problems/verify-preorder-serialization-of-a-binary-tree/


class VerifyPreorderSerializationofaBinaryTree(TestCase):
    def isValidSerialization(self, preorder):
        """
        :type preorder: str
        :rtype: bool
        """
        df = 1
        for s in preorder.split(","):
            if df == 0: return False
            df += -1 if s == "#" else 1
        return df == 0

    def test1(self):
        self.assertEqual(True, self.isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#"))

    def test2(self):
        self.assertEqual(False, self.isValidSerialization("1,#"))

    def test3(self):
        self.assertEqual(False, self.isValidSerialization("9,#,#,1,#,#"))

    def test4(self):
        self.assertEqual(False, self.isValidSerialization(""))

    def test5(self):
        self.assertEqual(True, self.isValidSerialization("#"))

File: kotlin/src/main/kotlin/331.VerifyPreorderSerializationofaBinaryTree.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/verify-preorder-serialization-of-a-binary-tree/


class VerifyPreorderSerializationofaBinaryTree{
  fun isValidSerialization(preorder: String): Boolean {
    val str = preorder.split(",")
    val n = str.size
    var i = 0
    fun dfs(): Boolean{
      if(i >= n )return false

      return if(str[i++] == "#")
        true
      else
        dfs() && dfs()
    }
    return dfs() && i == n
  }

  @Test
  fun test1(){
    assertEquals(true, isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#"))
  }

  @Test
  fun test2(){
    assertEquals(false, isValidSerialization("1,#"))
  }

  @Test
  fun test3(){
    assertEquals(false, isValidSerialization("9,#,#,1"))
  }

  @Test
  fun test4(){
    assertEquals(false, isValidSerialization(""))
  }

  @Test
  fun test5(){
    assertEquals(true, isValidSerialization("#"))
  }
}